Start: create thumbnails using f3d

chennes opened the pull request:
This is based on, and a replacement for, #17069 -- it extends the work in that PR to shift the processing into a set of workers running from the `QThreadPool` and returning their data as it arrives.

It also adds Start to the C++ test suite, and adds a couple of tests.

chennes commented:
@andesfreedesign my version of `f3d`, 2.2.1, does not understand the `--config=thumbnail` or `--verbose=quiet` arguments. Which version are you using in your tests?

chennes commented:
@adrianinsaval could you check this out and make sure it does not freeze up your system when it's generating a lot of previews? I'm just using `QThreadPool` without doing any extra work, but I can probably do some equivalent of `nice` to reduce the threads' priority if needed.

benj5378 commented on src/Mod/Start/App/DisplayedFilesModel.h:
```diff
@@ -28,6 +28,7 @@
 #include <Base/Parameter.h>
 
 #include "../StartGlobal.h"
+#include "ThumbnailSource.h"
```
```suggestion
```

Move to .cpp

benj5378 commented on src/Mod/Start/App/FileUtilities.h:
```diff
@@ -0,0 +1,55 @@
+#include "../StartGlobal.h"
+
+#include <string>
+#include <Base/FileInfo.h>
+#include <QString>
+#include <QDir>
```
```suggestion

class QString;
class QDir;
```
This is just a moved/renamed file, so leave this comment if you want

benj5378 commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,100 @@
+    }
+}
+
+QString Start::getMD5Hash(const std::string& path)
+{
+    // Use MD5 hash as specified here:
+    // https://specifications.freedesktop.org/thumbnail-spec/0.8.0/thumbsave.html
+    QUrl url(QString::fromStdString(path));
```
```suggestion
QString Start::getMD5Hash(const QString& path)
{
    // Use MD5 hash as specified here:
    // https://specifications.freedesktop.org/thumbnail-spec/0.8.0/thumbsave.html
    QUrl url(path);

```
Doing this, we can avoid converting to std::string the whole chain up/down

```cpp
QString Start::getMD5Hash(const std::string& path)
QString Start::getUniquePNG(const std::string& path)
const QString thumbnailPath = getUniquePNG(_file.toStdString());
```

benj5378 commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,100 @@
+    // Use MD5 hash as specified here:
+    // https://specifications.freedesktop.org/thumbnail-spec/0.8.0/thumbsave.html
+    QUrl url(QString::fromStdString(path));
+    url.setScheme(QString::fromLatin1("file"));
```
```suggestion
    url.setScheme(QStringLiteral("file"));

```

adrianinsaval commented:
Haven't tested big step files yet but this how the thumbnails show up for me on windows, just a logo
![image](https://github.com/user-attachments/assets/aabc8804-47cb-4221-9abd-00d3bbd71d4a)

chennes commented:
It didn't even occur to me that this might work on Windows! Definitely didn't test that.

chennes commented:
@benj5378 what am I supposed to be doing with all these `clazy-qstring-allocations` warnings?

adrianinsaval commented:
> Haven't tested big step files yet but this how the thumbnails show up for me on windows, just a logo

nevermind, it was just because f3d wasn't on path so wasn't actually being called, once I added the parameter with the binary Path it works but the filename is overlayed on top of the images 
![image](https://github.com/user-attachments/assets/026ed989-1199-417c-bb4c-d9d66eb47241)
I believe that was not the case in the previous PR, maybe because we're not using `--config=thumbnail` anymore

chennes commented:
Maybe - I had to remove that because on my Ubuntu 22.04 test system the installed version of `f3d` didn't support the option. Maybe we should be smarter about it, and detect the version?

benj5378 commented:
> @benj5378 what am I supposed to be doing with all these `clazy-qstring-allocations` warnings?

As far as I can tell from https://forum.freecad.org/viewtopic.php?t=65560 and https://github.com/KDE/clazy/blob/master/docs/checks/README-qstring-allocations.md , then the idea behind using `QLatin1String` is `QString` that have comparative overloads that are faster than QStringLiteral.

However, QString and QLatin1Strnig are different classes, so constructing a `QString` from `QLatin1String` means allocating.

[Clazy docs](https://github.com/KDE/clazy/blob/master/docs/checks/README-qstring-allocations.md):
```cpp
QString s = QStringLiteral("foo"); // No allocation
QString s = QLatin1String("foo"); // Allocates, use QStringLiteral("foo") instead
s == QLatin1String("foo") // No allocation
s == QStringLiteral("foo") // No allocation
```

[Qt docs](https://doc.qt.io/qt-5/qlatin1string.html#details):
> Note: If the function you're calling with a QLatin1String argument isn't actually overloaded to take QLatin1String, the implicit conversion to [QString](https://doc.qt.io/qt-5/qstring.html) will trigger a memory allocation, which is usually what you want to avoid by using QLatin1String in the first place. In those cases, using [QStringLiteral](https://doc.qt.io/qt-5/qstring.html#QStringLiteral) may be the better option.

So what should you do? You get the warning because you construct a QString from a QLatin1String, thus implicitly calling the constructor and allocating. Change all use of `QLatin1String` to `QStringLiteral`, (where we don't have an overload for `QLatin1String`, nearly all cases).

andesfreedesign commented:
> @andesfreedesign my version of `f3d`, 2.2.1, does not understand the `--config=thumbnail` or `--verbose=quiet` arguments. Which version are you using in your tests?

Sorry, but I was on vacation...
My F3D version is 2.5.0

adrianinsaval commented:
does `--filename=0` option work on your version @chennes ?

adrianinsaval commented:
`--filename=0 --grid=0 --no-background` gives a more similar result to `--config=thumbnail`, we should also set `--max-size` value, thumbnail config has it at 100Mib on my version, might make sense to make it even lower. For reference here's what I think is used when we use `--config=thumbnail` option

> cat .\thumbnail.d\05_all.json

```json
[
{
  "options":
  {
    "anti-aliasing": true,
    "camera-direction": "-1,-0.5,-1",
    "hdri-ambient": true,
    "max-size":100,
    "no-background": true,
    "verbose": "quiet",
    "tone-mapping": true,
    "translucency-support": true
  }
}
]
```

> cat .\thumbnail.d\10_occt.json

```json
[
{
  "match": ".*(step|stp|iges|igs|brep|xbf)",
  "options":
  {
    "scalar-coloring": true,
    "load-plugins": "occt",
    "up": "+Z",
    "ambient-occlusion": true,
    "coloring-component": "-2",
    "coloring-by-cells": true,
    "camera-direction": "-1,1,-0.5"
  }
}
]
```

while this is what is used normally:

> cat .\config.d\05_all.json
```json
[
{
  "options":
  {
    "axis": true,
    "tone-mapping": true,
    "grid": true,
    "loading-progress": true,
    "anti-aliasing": true,
    "filename": true,
    "camera-direction": "-1,-0.5,-1",
    "hdri-ambient": true,
    "translucency-support": true,
    "animation-progress": true
  }
}
]
```

> cat .\config.d\10_occt.json

```json
[
{
  "match": ".*(step|stp|iges|igs|brep|xbf)",
  "options":
  {
    "scalar-coloring": true,
    "load-plugins": "occt",
    "up": "+Z",
    "ambient-occlusion": true,
    "coloring-component": "-2",
    "coloring-by-cells": true,
    "camera-direction": "-1,1,-0.5"
  }
}
]
```
we could also consider enabling `--camera-orthographic` depending on user's view preference

chennes commented:
Thanks, @adrianinsaval -- can you add that as a GitHub `suggestion` so I can just adopt the commit?

adrianinsaval commented:
I just tested a 600mb file, freecad doesn't freeze but I eventually get this warning:
`17:28:52  QProcess: Destroyed while process ("C:\\Program Files\\F3D\\bin\\f3d.exe") is still running.`
the warning is shown after some time every time I start freecad or more specifically every time Start is loaded
setting max-size will likely greatly reduce how often users can encounter this warning but it might still a good idea to redirect those to log level or straight up ignore them

adrianinsaval commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,81 @@
+    const auto f3d = QString::fromStdString(_startParameterGrp->GetASCII("f3d", "f3d"));
+    constexpr int resolution = 128;
+    QStringList args;
+    args << QLatin1String("--load-plugins=occt") << QLatin1String("--output=") + thumbnailPath
```
```suggestion
    args << QLatin1String("--load-plugins=occt") << QLatin1String("--output=") + thumbnailPath
         << QLatin1String("--filename=0") << QLatin1String("--grid=0")
         << QLatin1String("--no-background") << QLatin1String("--max-size=100")
```

chennes commented:
`f3d` turns out to be something of a moving target: I just tried it on Ubuntu 22.04LTS and it's only version 1.2 there, which doesn't have a "--load-plugins" argument (for example). How much maintenance are we committing ourselves to here?

adrianinsaval commented:
> `f3d` turns out to be something of a moving target: I just tried it on Ubuntu 22.04LTS and it's only version 1.2 there, which doesn't have a "--load-plugins" argument (for example). How much maintenance are we committing ourselves to here?

what where you testing on before? I don't think we need to support every single feature in 22.04 IMO it's okay to ask for a higher minimum f3d version for this non critical QoL feature. So perhaps it makes sense to have a version check before attempting to run f3d. On the other hand I don't know if the `--load-plugins` argument is even required, I never needed it when calling f3d 2.5.0 from command line.

chennes commented:
> > `f3d` turns out to be something of a moving target: I just tried it on Ubuntu 22.04LTS and it's only version 1.2 there, which doesn't have a "--load-plugins" argument (for example). How much maintenance are we committing ourselves to here?
> 
> what where you testing on before? I don't think we need to support every single feature in 22.04 IMO it's okay to ask for a higher minimum f3d version for this non critical QoL feature. So perhaps it makes sense to have a version check before attempting to run f3d. On the other hand I don't know if the `--load-plugins` argument is even required, I never needed it when calling f3d 2.5.0 from command line.

I needed the `--load-plugins=occt` to generate previews for STEP files (which I'd guess is a pretty common use-case).

chennes commented:
OK: the documentation for pre-2.0 releases is pretty sketchy, and it looks like I can write support for 2.x and 3.x, so I will add a bit of code to do a version check, and cut off at 2.0 as the oldest we support. That was only two years ago, but this software is undergoing quite rapid development.

chennes commented:
@FreeCAD/code-quality-working-group could you please take a look at ThumbnailSource.*? Particularly to make sure I'm thread-safe there, but anything else that catches your eye as well. Thanks!

pieterhijma commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,102 @@
+    return QString::fromLatin1(ba);
+}
+
+QString Start::getUniquePNG(const QString& path)
```
```suggestion
QString Start::getPathCachedThumbnail(const QString& path)
```

pieterhijma commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,102 @@
+    return dir.absoluteFilePath(md5);
+}
+
+bool Start::useCachedPNG(const QString& image, const QString& project)
```
```
bool Start::useCachedThumbnail(const QString& pathThumbnail, const QString& pathProject)
```

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+
+using namespace Start;
+
+bool ThumbnailSource::_f3dInitialized = false;
```
These 4 static variables are governed by the mutex.  At the least I would add the comment that this is the case and that each thread must go through the mutex before reading any value, because otherwise a thread misses the barrier provided by the mutex and can read a stale value.

I would create a struct for these 4 variables to make more clear that the contents of the struct variables are governed by the mutex.  To me this would literally give more structure to the code.

You could go fancy and make a class where the getters and setters go through a lock (readers/writers lock would be appropriate in that case) and this would work in this example, but in general I'm against mixing synchronization with data structures, so a struct with the 4 variables and the mutex to govern access to it would be ideal from my point of view.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+void ThumbnailSource::run()
+{
+    const QString thumbnailPath = getUniquePNG(_file);
+    if (!useCachedPNG(thumbnailPath, _file)) {
```
I was a bit worried about contention on the mutex if there are many thumbnails, but since there is a check for a cached value, I think the contention is in general low, so it looks like the right synchronization granularity to me.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+{
+    const QString thumbnailPath = getUniquePNG(_file);
+    if (!useCachedPNG(thumbnailPath, _file)) {
+        setupF3D();
```
This is where the various threads access the mutex and where the memory is synchronized between them, after this, the 4 variables can safely be read. I think this warrants a comment because it is very important.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+        QStringList args(_f3dBaseArgs);
+        args << QLatin1String("--output=") + thumbnailPath << _file;
+
+        _process = std::make_unique<QProcess>();
```
There are two places where the process is started, both with slightly different arguments.  This one is busy waiting, the other one is blocking. Instead of this, I would use Qt's signalling system to get rid of blocking and busy waiting. I would also try to factor out the duplication and to make an abstraction to call the process.

chennes replied:
What is the advantage here? We ought to have the loop anyway to check for user interrupt requests, so it doesn't seem to provide any advantage to also have a slot to handle completion, and it adds complexity to a system that's by construction running in its own thread.

pieterhijma replied:
Good question. I dislike busy waiting in general. For example, there needs to be a decision on the granularity which is 50 ms here but there is no comment on whether this is a good value, so to me it is just a magic number.

Slots allow you to check for whether an error has occurred as well and it has the benefit that you don't have to come up with some number.

I studied the QProcess documentation a bit and indeed it says that in a situation like this (having its own thread), these calls allow you to simplify the code by just waiting.  So, I'm ok if we are choosing that option but then I would prefer a comment that explains why we are using blocking calls. However, I prefer the version without a busy-waiting loop that returns after its default value of 30 s after which you can check the result of the process. I think that is a good default value for a thumbnail creating thread.

To be able to do that, we need to do something with `isInterruptionRequested()` and I noticed that nowhere in this code it is actually requested, so, it can be removed from the loop as far as I can tell.

Regarding the refactoring, it minimizes code duplication.

Finally, I prefer this code to be well defined and well commented with regards to the design decisions, because it is very common in FreeCAD that code or patterns like these are copied to other places. Could it be the case that that has happened here as well with the `isInterruptionRequested()` pattern from the AddonManager in which `requestInterruption()` is actually called? Having this code well-defined in the first place helps that practice as well. Another argument in favor of having this code to be well defined and well commented is the fact that we're dealing with concurrency.

chennes replied:
An interrupt *should* be getting called when Start is closed, so that's what I'm coding for here. Is there another way to handle that interrupt request besides spooling up this type of event loop? (I have no objection to adding clarifying comments, of course).

chennes replied:
You are right: I coded for a requested interruption, but it didn't work the way I wanted it to, I ended up switching to just killing the subprocess and never circled back around to this loop. I'm happy to make the change to the signal/slot mechanism.

pieterhijma replied:
Hm, too bad we need two versions. Actually, I tried to explain that after reading the QProcess documentation, I actually liked the solution with a blocking call, so `waitForFinished()` because it has this timeout of 30 s. Since we don't need to check for interrupts, we don't need the busy-wait loop with an arbitrary value. I'm also ok with this code, but it has the drawback that if f3d misbehaves and for example enters an endless loop, it will run forever until the application is killed.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+        Base::Console().Log("Creating thumbnail for %s...\n", _file.toStdString());
+        _process->start(f3d, args);
+        constexpr int checkEveryMs {50};
+        while (!_process->waitForFinished(checkEveryMs)) {
```
The busy waiting

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+        return;
+    }
+
+    _f3dInitialized = true;  // Set immediately so we can use early-return below
```
Counterintuitive to set the flag to true here. The comment helps.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+        "User parameter:BaseApp/Preferences/Mod/Start");
+    const auto f3d = QString::fromUtf8(hGrp->GetASCII("f3d", "f3d").c_str());
+    const QStringList args {QLatin1String("--version")};
+    QProcess process;
```
The other place where the process is started.

chennes replied:
I've added an explanatory comment to the main mutex-protected method to communicate the design decision made here. I believe that blocking is the best thing to do in this initialization method, and no other approach really makes sense.

pieterhijma replied:
Yes, I agree, see my comment above about blocking.

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,187 @@
+    const QStringList args {QLatin1String("--version")};
+    QProcess process;
+    process.start(f3d, args);
+    process.waitForFinished();
```
blocking here.

chennes replied:
I'm blocking here because there's no realistic probability of a user-requested interruption of a call that is only requesting the version output from `f3d` (it's not asking the software to do a conversion here).

pieterhijma commented on tests/src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
```
With these tests, we can also include the tests for the human readable file sizes. From the forum:

```cpp
#include "gtest/gtest.h"

#include "Mod/Start/App/DisplayedFilesModel.h"

using namespace Start;

TEST(StartTests, B)
{
    EXPECT_EQ("999 B", DisplayedFilesModel::humanReadableSize(999));
}

TEST(StartTests, kB)
{
    EXPECT_EQ("1.2 kB", DisplayedFilesModel::humanReadableSize(1234));
}

TEST(StartTests, kB)
{
    EXPECT_EQ("1.0 kB", DisplayedFilesModel::humanReadableSize(1000));
}

TEST(StartTests, kB_eq)
{
    EXPECT_EQ("1.0 kB", DisplayedFilesModel::humanReadableSize(1001));
}

TEST(StartTests, MB)
{
    EXPECT_EQ("2.5 MB", DisplayedFilesModel::humanReadableSize(2.5e6));
}

TEST(StartTests, GB)
{
    EXPECT_EQ("3.7 GB", DisplayedFilesModel::humanReadableSize(3.7e9));
}

TEST(StartTests, TB)
{
    EXPECT_EQ("444.8 TB", DisplayedFilesModel::humanReadableSize(444.823e12));
}

TEST(StartTests, PB)
{
    EXPECT_EQ("1.2 PB", DisplayedFilesModel::humanReadableSize(1.2e15));
}

TEST(StartTests, EB)
{
    EXPECT_EQ("7.3 EB", DisplayedFilesModel::humanReadableSize(7.3e18));
}
```

pieterhijma commented on src/Mod/Start/App/DisplayedFilesModel.cpp:
```diff
@@ -26,28 +26,31 @@
 {
-
-std::string humanReadableSize(unsigned int bytes)
+std::string humanReadableSize(const unsigned int bytes)
```
Actually, it is totally true that you can compute it at once and yes I agree that that is a much better implementation.

Well, as I said, it is a bit in scope of this implementation but I see your point.

hyarion replied:
I reread my comment and it was a bit harsh. Sorry.
What I meant was that I think it is a bigger discussion, which we tried before and failed come to a conclusion on (developer handbook c++ guidelines). I don't think it is fair to the authors to bring it into this PR as I believe it would stall this PR.

I don't see that it is needed here, so that was why I thought it would be better to discuss it elsewhere. But I might just have missed the reason :)

hyarion replied:
Regarding removing the anonymous namespace. If that is done, could you place the function to the `Start` namespace to avoid polluting the global?

chennes replied:
Just for the record, switching to using a log10 calculation is a clearly superior solution here, so I am planning on making that change.

pieterhijma replied:
> I reread my comment and it was a bit harsh. Sorry. What I meant was that I think it is a bigger discussion, which we tried before and failed come to a conclusion on (developer handbook c++ guidelines). I don't think it is fair to the authors to bring it into this PR as I believe it would stall this PR.

No worries, your points were very valid.

chennes replied:
> Regarding removing the anonymous namespace. If that is done, could you place the function to the `Start` namespace to avoid polluting the global?

Yes, this happened when I moved it to FileUtilities.h and refactored it to use a log10 calculation.

chennes commented:
Thanks for the review, @pieterhijma -- I'll start making these changes.

chennes commented on src/Mod/Start/App/DisplayedFilesModel.cpp:
```diff
@@ -26,28 +26,31 @@
 using namespace Start;
 
 
 namespace
```
I've moved that method into `FileUtilities.h` and added unit tests.

chennes commented on src/Mod/Start/App/DisplayedFilesModel.cpp:
```diff
@@ -303,16 +261,23 @@ void DisplayedFilesModel::clear()
-        std::make_pair(int(DisplayedFilesModelRoles::modifiedTime), "modifiedTime"),
-        std::make_pair(int(DisplayedFilesModelRoles::path), "path"),
-        std::make_pair(int(DisplayedFilesModelRoles::size), "size"),
+        std::make_pair(static_cast<int>(DisplayedFilesModelRoles::author), "author"),
```
I'm going to relegate this to "later work" since it's unrelated to this PR.

pieterhijma replied:
Yes ok, I hoped that I could get rid of the `make_pair` as well, but since that doesn't work, you could argue that my solution adds even more boilerplate.

chennes commented:
@pieterhijma I've addressed the majority of your comments above -- exceptions are noted inline. I'm pushing a new commit so you can take a look at the changes, though I'm having some difficulty getting the test code to link on my local build, so I expect it to also fail in CI here.

adrianinsaval commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,213 @@
+        QStringLiteral("--verbose=quiet"),
+        QStringLiteral("--resolution=256,256"),
+        QStringLiteral("--filename=0"),
+        QStringLiteral("--grid=0"),
```
```suggestion
        QStringLiteral("--grid=0"),
        QStringLiteral("--axis=0"),
```
was missing from my previous suggestion

pieterhijma commented on src/Mod/Start/App/DisplayedFilesModel.cpp:
Nice, good improvement from my perspective!

pieterhijma commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    const double logBaseFactor = std::log10(unitFactor);  // change to constexpr on c++26
+    const auto unitIndex = static_cast<size_t>(std::log10(bytes) / logBaseFactor);
+    const auto unit = units.at(unitIndex);
```
Nice! This is indeed a better implementation than the one I suggested from the forum. The comment on line 84 implies it but I would like an explicit comment here that `unitIndex` cannot go out of range in `units` because uint64_t can't express numbers higher than the EB range. When I read the code, this was immediately a question that popped up and having a simple comment here can stop a whole thought process for someone reading the code :)

chennes replied:
Done.

pieterhijma commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    const bool isByteUnit = unitIndex == 0;
+    const size_t precision = isByteUnit ? 0 : 1;
+    return fmt::format("{:.{}f} {}", scaledValue, precision, unit);  // change to std:: on c++20
```
We are now on c++20, right?

chennes replied:
Yes, but the comment is wrong: we are currently still using libfmt. Removed.

pieterhijma commented on tests/src/Mod/Start/App/FileUtilities.cpp:
Great to have these tests.

pieterhijma reviewed (changes requested):
Thanks for making the changes! Looks good to me. I just have a couple of small things, see the inline comments.

chennes commented:
@pieterhijma your comment about the timeout reminded me that we don't actually properly handle that case, so I've added the code to do so (the timeout doesn't actually kill the process, but the code assumed that after the `wait` the process had completed, which is not necessarily true). I've also added a timeout to stop the job after 30 seconds so it doesn't just consume this thread forever.

hyarion commented on src/Mod/Start/App/FileUtilities.cpp:
```diff
@@ -0,0 +1,109 @@
+
+    // uint64_t can't express numbers higher than the EB range (< 20 EB)
+    constexpr std::array units {"B", "kB", "MB", "GB", "TB", "PB", "EB"};
+    constexpr double unitFactor = 1000.0;
```
Remember this comment? https://github.com/FreeCAD/FreeCAD/pull/17817#issuecomment-2471042167

3x380V replied:
Not sure how does it answer the above question, so again: why should FreeCAD display different numbers (for sizes up to 1000B numbers would be the same) than each and every file manager ever used?

hyarion replied:
On your question
> Is there particular reason for this behaviour?

The comment answers why it was done like that before
> I originally went with the 1000 divisor because I personally prefer the MB to MiB etc. -- the MiB always felt overly pedantic to me (and they are awkward to pronounce :) ). But that is pure personal preference, I'm not going to hold this PR up for that.

I don't think anyone would mind much if you like it to be 1024 and still use MB etc. instead - like in Finder etc.

3x380V replied:
It is not what I would like, it is common practice. FreeCAD is just showing different size than everything else. I'm not aware about any software using 1000 base. To make it clear, I would of course like it to be denoted according to IEC 80000-13 (kiB, MiB...), just for a fun of educating users, but using kB for 1024, MB for 1024×1024, etc. as semiconductor memory industry does for ages seems appropriate. The key point is to avoid question like why does FreeCAD shows different file size than my file manager.

hyarion replied:
It looks like I'm wrong, sorry.

Finder does use 1000, not 1024 (since 10.8). The same thing applies for all software on mac which uses the foundation framework for converting number of bytes to a string and doesn't explicitly specify to use 1024: https://developer.apple.com/documentation/foundation/bytecountformatter/countstyle (it always uses MB never MiB)

According to IEC 80000-13, MB and KB is valid to use when using 1000 as factor. It does specify that if 1024 is used, it should be named KiB, MiB etc. (look at the examples).

By your arguments (follow standard + consistency with os file explorer), I suppose the current version actually is preferred on macOS?

3x380V replied:
Ouch! I wasn't aware Apple changed to SI standard in 10.8. Anyway, IEC 80000-13 states in the second paragraph _The following referenced documents are indispensable for the application of this document_ but, duh!, referenced ISO-IEC-2382-16 is no longer valid and I have not purchased current norm.

So indeed, current implementation matches current MacOS.

hyarion replied:
https://www.iso.org/standard/7259.html has been replaced by https://www.iso.org/standard/63598.html which is available for free ($0 but need to go though the checkout process) or accessed from the online platform: https://www.iso.org/obp/ui/en/#!iso:std:63598:en

chennes commented:
Sorry for the delay on this, it turned out that I had an error in the logic about when the new thumbnail signal was emitted. I fixed that, and also added the code to correctly emit the `dataChanged()` signal from the model when the new thumbnails become available, so that they appear as soon as possible without needing any sort of user-driven refresh.

mwestphal commented:
Hi, F3D maintainers here! Do not hesitate to ping me if you need explanation about our options or our doc :)

oursland commented:
The f3d gallery is pretty impressive!  OCCT is in the process of adding support to representing animations in their data structures (https://github.com/Open-Cascade-SAS/OCCT/pull/492), which will eventually feed into being able to export animations via glTF or other formats.  I can see further integration of f3d to generate renders of assemblies, exploded parts, and animations.

maxwxyz commented:
^ @FEA-eng @marioalexis84 heads up for FEM animations and @mfasano727

chennes commented:
@pieterhijma maybe use the "suggestion" feature here to show what you mean, and I can just merge it in?

pieterhijma commented on src/Mod/Start/App/ThumbnailSource.cpp:
```diff
@@ -0,0 +1,190 @@
+{
+    _thumbnailPath = getPathToCachedThumbnail(_file);
+    if (!useCachedThumbnail(_thumbnailPath, _file)) {
+        setupF3D();  // Go through the mutex to ensure data is not stale
```
``` suggestion
        // Go through the mutex to ensure data is not stale.
        // Contention on the lock is diminished because of first checking the cache.
        setupF3D();
```

pieterhijma replied:
Good idea. Something like this?
